[examples] add train flux-controlnet scripts in example. (#9324)
huggingface/diffusers: examples/controlnet/README_flux.md

# ControlNet training example for FLUX

The `train_controlnet_flux.py` script shows how to implement the ControlNet training procedure and adapt it for [FLUX](https://github.com/black-forest-labs/flux).

> Training script provided by LibAI, which is an institution dedicated to the progress and achievement of artificial general intelligence.LibAI is the developer of [cutout.pro](https://www.cutout.pro/) and [promeai.pro](https://www.promeai.pro/).

Reviewer: ```suggestion
Training script provided by LibAI, which is an institution dedicated to the progress and achievement of artificial general intelligence. LibAI is the developer of [cutout.pro](https://www.cutout.pro/) and [promeai.pro](https://www.promeai.pro/).
```

    --pretrained_model_name_or_path=$MODEL_DIR \
    --conditioning_image_column=control_path \
    --image_column=image_path \
    --caption_column=caption \
    --output_dir=$OUTPUT_DIR \
    --jsonl_for_train=$TRAIN_JSON_FILE \
    --mixed_precision="bf16" \
    --resolution=512 \
    --learning_rate=1e-5 \
    --max_train_steps=15000 \
    --validation_steps=100 \
    --checkpointing_steps=200 \
    --validation_image "./conditioning_image_1.png" "./conditioning_image_2.png" \
    --validation_prompt "red circle with blue background" "cyan circle with brown floral background" \
    --train_batch_size=1 \
    --gradient_accumulation_steps=4 \
    --report_to="tensorboard" \
    --num_double_layers=4 \
    --num_single_layers=0 \
    --seed=42 \
```

To better track our training experiments, we're using the following flags in the command above:

* `report_to="tensorboard` will ensure the training runs are tracked on Weights and Biases.
* `validation_image`, `validation_prompt`, and `validation_steps` to allow the script to do a few validation inference runs. This allows us to qualitatively check if the training is progressing as expected.

> Our experiments were conducted on a single 40GB A100 GPU.

Reviewer: Wow, 40GB A100 seems doable.

Author: I'm sorry, this is the 80g A100 (I wrote it wrong), I did a lot of extra work to get it to train with the zero3 on the 40g A100, but I don't think this is suitable for everyone

Reviewer: Not at all. I think it would still be nice to include the changes you had to make in the form of notes in the README. Does that work?

Author: I'll see if I can add it later.

Author: @sayakpaul We added a tutorial on configuring deepspeed in the readme.

Reviewer 2: There are some tricks to lower GPU:
1) gradient_checkpointing
2) bf16 or fp16. 
3) batch size 1, and then use gradient_accumulation_steps above 1

With 1, 2, 3, can this thing be controlled to be trained under 40GB?

Author: According to my practice, deepspeedzero3 must be used

Reviewer 3: sorry to bother you, have you ever tried cache text-encoder and vae latents to run with lower GPU？ @PromeAIpro @linjiapro

Author: cache text-encoder is already available in this script (saving about 10g of gpu memory on T5), about cache vae You can check how to use deepspeed in the readme, which includes cache vae.

### Inference

Once training is done, we can perform inference like so:

```python
import torch
from diffusers.utils import load_image
from diffusers.pipelines.flux.pipeline_flux_controlnet import FluxControlNetPipeline
from diffusers.models.controlnet_flux import FluxControlNetModel

> base_model = 'black-forest-labs/FLUX.1-dev'
> controlnet_model = 'path to controlnet'
> controlnet = FluxControlNetModel.from_pretrained(controlnet_model, torch_dtype=torch.bfloat16)
> pipe = FluxControlNetPipeline.from_pretrained(base_model, 
>                                               controlnet=controlnet, 
>                                               torch_dtype=torch.bfloat16)
> pipe.to("cuda")

Reviewer: ```suggestion
base_model = 'black-forest-labs/FLUX.1-dev'
controlnet_model = 'path to controlnet'
controlnet = FluxControlNetModel.from_pretrained(controlnet_model, torch_dtype=torch.bfloat16)
pipe = FluxControlNetPipeline.from_pretrained(
    base_model, 
    controlnet=controlnet, 
    torch_dtype=torch.bfloat16
)
# enable memory optimizations   
pipe.enable_model_cpu_offload()
```

Most people may not have the necessary VRAM to run it like this. So, better have it this way? WDYT?

Author: yes, you are right

control_image = load_image("./conditioning_image_1.png").resize((1024, 1024))
prompt = "pale golden rod circle with old lace background"

image = pipe(
    prompt, 
    control_image=control_image,
    controlnet_conditioning_scale=0.6,
    num_inference_steps=28, 
    guidance_scale=3.5,
).images[0]
image.save("./output.png")
```

## Notes

### T5 dont support bf16 autocast and i dont know why, will cause black image.

```diff
if is_final_validation or torch.backends.mps.is_available():
    autocast_ctx = nullcontext()
else:
    # t5 seems not support autocast and i don't know why
+   autocast_ctx = nullcontext()
-   autocast_ctx = torch.autocast(accelerator.device.type)

> ```

Reviewer: Instead of this how about we directly incorporate the fix in the training script itself?

Author: done, plz check it.

> ### TO Fix Error

Reviewer: Same, let's fix this in the pipeline implementation itself.

Author: done, plz check it.

# ControlNet training example for FLUX

The `train_controlnet_flux.py` script shows how to implement the ControlNet training procedure and adapt it for [FLUX](https://github.com/black-forest-labs/flux).

Training script provided by LibAI, which is an institution dedicated to the progress and achievement of artificial general intelligence. LibAI is the developer of [cutout.pro](https://www.cutout.pro/) and [promeai.pro](https://www.promeai.pro/).

> 

Reviewer: ```suggestion
> [!NOTE]
> **Memory consumption**
>
> Flux can be quite expensive to run on consumer hardware devices and as a result, ControlNet training of it comes with higher memory requirements than usual. 

> **Gated access**
> As the model is gated, before using it with diffusers you first need to go to the [FLUX.1 [dev] Hugging Face page](https://huggingface.co/black-forest-labs/FLUX.1-dev), fill in the form and accept the gate. Once you are in, you need to log in so that your system knows you’ve accepted the gate. Use the command below to log in: `huggingface-cli login`

```

    --checkpointing_steps=200 \
    --validation_image "./conditioning_image_1.png" "./conditioning_image_2.png" \
    --validation_prompt "red circle with blue background" "cyan circle with brown floral background" \
    --train_batch_size=1 \
    --gradient_accumulation_steps=4 \
    --report_to="wandb" \
    --num_double_layers=4 \
    --num_single_layers=0 \
    --seed=42 \
    --push_to_hub \
```

To better track our training experiments, we're using the following flags in the command above:

* `report_to="wandb` will ensure the training runs are tracked on Weights and Biases.
* `validation_image`, `validation_prompt`, and `validation_steps` to allow the script to do a few validation inference runs. This allows us to qualitatively check if the training is progressing as expected.

Our experiments were conducted on a single 80GB A100 GPU.

### Inference

Once training is done, we can perform inference like so:

```python
import torch
from diffusers.utils import load_image
from diffusers.pipelines.flux.pipeline_flux_controlnet import FluxControlNetPipeline
from diffusers.models.controlnet_flux import FluxControlNetModel

base_model = 'black-forest-labs/FLUX.1-dev'

> controlnet_model = 'path to controlnet'

Reviewer: Let's provide the repo id of an already trained model (on our conditioning dataset) here if possible?

Reviewer 4: I don't have such a model yet, maybe consider providing 'InstantX/FLUX.1-dev-Controlnet-Canny'

Reviewer: But you ran experiments with the script as I understand it right? Could we not provide something that worked in those experiments?

Author: I don't have enough GPU resources to train and test on our conditional dataset. Our organization is open sourcing a dev_lineart controlnet model(The example we proposed above) trained on this script. Is that ok?

Reviewer: Of course that would be okay!

Reviewer: @PromeAIpro let's update this then?

Author: already open sourcing here:https://huggingface.co/promeai/FLUX.1-controlnet-lineart-promeai

write_basic_config()
```

When running `accelerate config`, if we specify torch compile mode to True there can be dramatic speedups.

## Custom Datasets

We support importing data from jsonl(xxx.jsonl),here is a brief example:
```sh
{"image_path": "xxx", "caption": "xxx", "control_path": "xxx"}
{"image_path": "xxx", "caption": "xxx", "control_path": "xxx"}
```

## Training

Our training examples use two test conditioning images. They can be downloaded by running

```sh
wget https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/controlnet_training/conditioning_image_1.png
wget https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/controlnet_training/conditioning_image_2.png
```

Then run `huggingface-cli login` to log into your Hugging Face account. This is needed to be able to push the trained ControlNet parameters to Hugging Face Hub.

we can define the num_layers, num_single_layers, which determines the size of the control(default values are num_layers=4, num_single_layers=10)

```bash
export MODEL_DIR="black-forest-labs/FLUX.1-dev"
export OUTPUT_DIR="path to save model"

> export TRAIN_JSON_FILE="path to your jsonl file"

Author: added in this [commit](https://github.com/huggingface/diffusers/pull/9324/commits/67deb7a696a850988f34e06aede7e0d9cc891d67) Plz take a look at it.

> export TRAIN_JSON_FILE="path to your jsonl file"

Reviewer: We should then change this command accordingly to reflect it no? Specifically, we should remove `export TRAIN_JSON_FILE="path to your jsonl file"`.

```bash
accelerate config
```

Or for a default accelerate configuration without answering questions about your environment

```bash
accelerate config default
```

Or if your environment doesn't support an interactive shell (e.g., a notebook)

```python
from accelerate.utils import write_basic_config
write_basic_config()
```

When running `accelerate config`, if we specify torch compile mode to True there can be dramatic speedups.

## Custom Datasets

We support dataset formats:
The original dataset is hosted in the [ControlNet repo](https://huggingface.co/lllyasviel/ControlNet/blob/main/training/fill50k.zip). We re-uploaded it to be compatible with `datasets` [here](https://huggingface.co/datasets/fusing/fill50k). Note that `datasets` handles dataloading within the training script, To use our example, add `--dataset_name=fusing/fill50k \` to the script and remove line `--jsonl_for_train` mentioned below.

We also support importing data from jsonl(xxx.jsonl),using `--jsonl_for_train` to enable it, here is a brief example of jsonl files:
```sh
{"image": "xxx", "text": "xxx", "conditioning_image": "xxx"}
{"image": "xxx", "text": "xxx", "conditioning_image": "xxx"}
```

> 

Reviewer: We can remove this whitespace.

Reviewer 4: fixed

```

Then run `huggingface-cli login` to log into your Hugging Face account. This is needed to be able to push the trained ControlNet parameters to Hugging Face Hub.

we can define the num_layers, num_single_layers, which determines the size of the control(default values are num_layers=4, num_single_layers=10)

```bash
export MODEL_DIR="black-forest-labs/FLUX.1-dev"
export OUTPUT_DIR="path to save model"
export TRAIN_JSON_FILE="path to your jsonl file"

accelerate launch train_controlnet_flux.py \
    --pretrained_model_name_or_path=$MODEL_DIR \
    --dataset_name=fusing/fill50k \
    --conditioning_image_column=conditioning_image \
    --image_column=image \
    --caption_column=text \
    --output_dir=$OUTPUT_DIR \
    --mixed_precision="bf16" \
    --resolution=512 \
    --learning_rate=1e-5 \
    --max_train_steps=15000 \
    --validation_steps=100 \
    --checkpointing_steps=200 \
    --validation_image "./conditioning_image_1.png" "./conditioning_image_2.png" \
    --validation_prompt "red circle with blue background" "cyan circle with brown floral background" \
    --train_batch_size=1 \
    --gradient_accumulation_steps=4 \

>     --report_to="tensorboard" \

Reviewer: We usually suggest logging to `wandb`.

Reviewer 4: fixed

    --num_double_layers=4 \
    --num_single_layers=0 \

>     --seed=42 \

Reviewer: ```suggestion
    --seed=42 \
    --push_to_hub
```

We usually default to `push_to_hub` in our examples.

Reviewer 4: fixed

```

To better track our training experiments, we're using the following flags in the command above:

> * `report_to="tensorboard` will ensure the training runs are tracked on Weights and Biases.

Reviewer: No, `report_to="wandb"` will ensure that.

Reviewer 4: fixed

* `validation_image`, `validation_prompt`, and `validation_steps` to allow the script to do a few validation inference runs. This allows us to qualitatively check if the training is progressing as expected.

Our experiments were conducted on a single 80GB A100 GPU.

### Inference

Once training is done, we can perform inference like so:

```python
import torch
from diffusers.utils import load_image
from diffusers.pipelines.flux.pipeline_flux_controlnet import FluxControlNetPipeline
from diffusers.models.controlnet_flux import FluxControlNetModel

base_model = 'black-forest-labs/FLUX.1-dev'
controlnet_model = 'promeai/FLUX.1-controlnet-lineart-promeai'
controlnet = FluxControlNetModel.from_pretrained(controlnet_model, torch_dtype=torch.bfloat16)
pipe = FluxControlNetPipeline.from_pretrained(
    base_model, 
    controlnet=controlnet, 
    torch_dtype=torch.bfloat16
)
# enable memory optimizations   
pipe.enable_model_cpu_offload()

> control_image = load_image("./conditioning_image_1.png").resize((1024, 1024))

Reviewer: Should we not change the conditioning image because we're using "promeai/FLUX.1-controlnet-lineart-promeai"?

Author: In fact, each image can be used as a test, but for demonstration purposes, I have modified it.
